PHPLIB-1163 Create tutorial for using MongoDB with Bref (#1273)
mongodb/mongo-php-library: docs/tutorial/aws-lambda.txt

==============================
Deploy to AWS Lambda with Bref
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

`Bref <https://bref.sh>`__ allows to deploy serverless PHP applications on AWS Lambda.
In this tutorial, you will deploy a simple PHP application with the MongoDB PHP extension,
and connect to an Atlas cluster using AWS IAM authentication.

Prerequisites
--------------

Before you begin, you must install Bref on your machine. You can follow the

> `official documentation to setup Bref <https://bref.sh/docs/setup>`__.

Author: Updated, with 2 items in the list (AWS, and Serverless)

connects to a MongoDB cluster.

.. code-block:: bash

    mkdir bref-mongodb-app && cd bref-mongodb-app
    composer init
    composer require bref/bref bref/extra-php-extensions mongodb/mongodb
    vendor/bin/bref init

The file ``index.php`` has been created. To validate the deployment, you can start
by deploying this default application.

.. code-block:: bash

    serverless deploy

Bref provides a Lambda layer with PHP and some very common extensions.
Additional extensions are provided by the package `bref/extra-php-extension <https://github.com/brefphp/extra-php-extensions>`__.

.. code-block:: yaml

    plugins:
        - ./vendor/bref/bref
        - ./vendor/bref/extra-php-extensions

    functions:
        api:
            handler: index.php
            runtime: php-83-fpm
            layers:
                - ${bref-extra:mongodb-php-83}

Let's try to use the MongoDB driver with this simple web page that list planets
from the `sample dataset <https://www.mongodb.com/docs/atlas/sample-data/>`__.
Replace the contents of ``index.php`` with the following:

.. literalinclude:: /examples/aws-lambda/index.php
   :language: php

> Deploy the application

Author: @mongodb/dbx-php What's your opinion on the formatting of this tutorial? It's actually steps.

> Deploy the application

Reviewer: No objections if you feel it's worth the effort.

Please make sure that formatting doesn't clash with "AWS Credentials", which has a numbered list within. I'm not expecting any issue nesting a list within the RST step syntax -- just a request to confirm the rendered output looks presentable.

> Deploy the application

Reviewer 2: I think skipping steps formatting is fine and more efficient with time. The issue I was focusing on was missing punctuation. Perhaps it could be something to experiment with in a future PR!

> Deploy the application

Author: Atlas links doesn't work. 
<img width="750" alt="image" src="https://github.com/mongodb/mongo-php-library/assets/400034/a75bc581-33df-48ad-a998-ff81bc9b4828">

.. code-block:: bash

    serverless deploy

> The application will not work unless you define the ``MONGODB_URI`` environment variable.

Reviewer 2: Question:

I wasn't able to determine whether you had addressed this -- if I missed it, can you point me to the change?

> The application will not work unless you define the ``MONGODB_URI`` environment variable.

Reviewer 2: > I propose this text:

Sure, I think that makes it more clear that it's describing the result of the run and the reason why.

AWS Credentials
---------------

Atlas supports passwordless authentication with AWS credentials. In any Lambda function,
AWS sets environment variables that contains the access token and secret token with
the role assigned to deployed function.

Set up `unified AWS Access <https://www.mongodb.com/docs/atlas/security/set-up-unified-aws-access/>`__:

1. Open the Lambda function in the AWS console
2. In "Configuration > Permission", copy the "Role name"
3. Open the MongoDB Atlas project
4. Go to "Security > Database Access"
5. Click "Add a new Database User"
6. Select Authentication Method: "AWS IAM", type "IAM Role" and paste the role name in "AWS Role ARN".
7. Add "Built-in Role": "Read and write any database"

> 8. Validate by clicking on "Add user".

Author: I have removed part of the procedure to refer to the Atlas doc page.

==============================
Deploy to AWS Lambda with Bref
==============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> .. versionadded:: 1.17

Author: Oups, a copy-paste leftover. I don't know the minimum version required. And I must confess that I don't really want to test all the old versions. The latest bref layer works.

>    :backlinks: none
>    :depth: 2
>    :class: singlecol
> 
> .. versionadded:: 1.17

Author: ```suggestion
   :backlinks: none
   :depth: 2
   :class: singlecol
```

Overview
--------

`Bref <https://bref.sh>`_ allows to deploy serverless PHP applications on AWS Lambda.
In this tutorial, you will deploy a simple PHP application with the MongoDB PHP extension,
and connect to an Atlas cluster using AWS IAM authentication.

Prerequisites
--------------

Before you begin, you must install Bref on your machine. You can follow the
`official documentation to setup Bref <https://bref.sh/docs/setup>`_.

Install the MongoDB extension
-----------------------------

By default, the bref layer is compiled with PHP and a few extensions. Additional extensions
are provided in additional layers.

Start by creating a new directory for your project and install the required MongoDB
and Bref dependencies. This project will be a bare minimum PHP web application that
connects to a MongoDB cluster.

.. code-block:: bash

    mkdir bref-mongodb-app && cd bref-mongodb-app
    composer init

>     composer require bref/bref bref/extra-php-extensions

Author: MongoDB package is missing.
```suggestion
    composer require bref/bref bref/extra-php-extensions mongodb/mongodb
```

    vendor/bin/bref init

The file ``index.php`` has been created. To validate the deployment, you can start
by deploying this default application.

.. code-block:: bash

    serverless deploy

Bref provides a Lambda layer with PHP and some very common extensions.

> Additional extensions are provided by the package `bref/extra-php-extension <https://github.com/brefphp/extra-php-extensions>`_.

Author: Done. Thanks for noticing this little detail.

.. code-block:: yaml

    plugins:
        - ./vendor/bref/bref
        - ./vendor/bref/extra-php-extensions

    functions:
        api:
            handler: index.php
            runtime: php-83-fpm
            layers:
                - ${bref-extra:mongodb-php-83}

Let's try to use the MongoDB driver with this simple web page that list planets

> from the :manual:`sample dataset </atlas/sample-data/>`.

Author: Indeed, I got it wrong. The prefix should be added. Maybe added to this standing PR: https://github.com/mongodb/docs-php-library/pull/57

> from the :manual:`sample dataset </atlas/sample-data/>`.

Author: Reverted to full url links for the time being.

> from the :manual:`sample dataset </atlas/sample-data/>`.

Reviewer 2: I think the extlinks entry may override the [Snooty role](https://github.com/mongodb/snooty-parser/blob/main/snooty/rstspec.toml#L1358-L1360).

For example, `docs-java` doesn't have a `conf.py` file and therefore no extlinks, but uses the `:manual:` role:
https://github.com/mongodb/docs-java/blob/612975d4e0a2b550a9eb486b75558afb6ee7c55b/source/usage-examples/watch.txt#L181-L183

> from the :manual:`sample dataset </atlas/sample-data/>`.

Reviewer: Thanks, I asked about this in [Slack](https://mongodb.slack.com/archives/C018UTGP23T/p1712930810725869) and will revise https://github.com/mongodb/docs-php-library/pull/57 based on the response.

Replace the contents of ``index.php`` with the following:

.. literalinclude:: /examples/aws-lambda/index.php
   :language: php

Deploy the application

.. code-block:: bash

    serverless deploy

The application will not work unless you define the ``MONGODB_URI`` environment variable.

AWS Credentials
---------------

Atlas supports passwordless authentication with AWS credentials. In any Lambda function,
AWS sets environment variables that contains the access token and secret token with
the role assigned to deployed function.

> Set up :manual:`unified AWS Access </atlas/security/set-up-unified-aws-access/>`:

Reviewer 2: I think you can use the `:atlas:` role unless it's overridden by extlinks:

https://github.com/mongodb/snooty-parser/blob/main/snooty/rstspec.toml#L1290-L1292

> Set up :manual:`unified AWS Access </atlas/security/set-up-unified-aws-access/>`:

Author: Atlas links doesn't work. 
<img width="750" alt="image" src="https://github.com/mongodb/mongo-php-library/assets/400034/a75bc581-33df-48ad-a998-ff81bc9b4828">

> Set up :manual:`unified AWS Access </atlas/security/set-up-unified-aws-access/>`:

Reviewer: > I think you can use the :atlas: role unless it's overridden by extlinks:

Does the very existence of `extlinks` in `conf.py` prevent us from using any of the roles defined in [main/snooty/rstspec.toml](https://github.com/mongodb/snooty-parser/blob/main/snooty/rstspec.toml)?

https://github.com/mongodb/docs-php-library/pull/57 was going to introduce `atlas` to `extlinks`, but it's presently not there.

As I mentioned in [Slack](https://mongodb.slack.com/archives/C018UTGP23T/p1712930810725869), we'd need a custom role for `:php:` to link to PHP.net pages. I'm not a bit worried that reducing `conf.py` to

```
extlinks = {
    'php': ('https://php.net/%s', '')
}
```

...might mean we lose access to `:manual:` as well.

> Set up :manual:`unified AWS Access </atlas/security/set-up-unified-aws-access/>`:

Reviewer: Disregard my last comment. @schmalliso [confirmed](https://mongodb.slack.com/archives/C018UTGP23T/p1712931118620829?thread_ts=1712930810.725869&cid=C018UTGP23T) that `conf.py` is ignored by Snooty. All of the roles we need are already defined in [main/snooty/rstspec.toml](https://github.com/mongodb/snooty-parser/blob/main/snooty/rstspec.toml) (including `:php:`).

@GromNaN: I'm not sure why `:atlas:` isn't working. Perhaps there's some other configuration where specific roles from `rstspec.toml` need to be enabled for a project? This seems like a question for `#ask-docs-platform`.

1. Open the Lambda function in the AWS console
2. In "Configuration > Permission", copy the "Role name"
3. Open the MongoDB Atlas project
4. Go to "Security > Database Access"
5. Click "Add a new Database User"
6. Select Authentication Method: "AWS IAM", type "IAM Role" and paste the role name in "AWS Role ARN".
7. Add "Built-in Role": "Read and write any database"
8. Validate by clicking on "Add user".

Now that the permissions have been configured, the lambda function is allowed to access
your Atlas cluster. You can configure your application with the Atlas endpoint.

Update the ``serverless.yml`` file to pass the environment variable ``MONGODB_URI``

.. code-block:: yaml

    provider:
        environment:
            MONGODB_URI: "mongodb+srv://cluster0.example.mongodb.net/"

The value can be found in "Atlas > Deployment > Database > Connect". Select "3. AWS IAM".
Remove the ``<AWS access key>:<AWS secret key>`` part from the URI, the credentials

> will be read from environment variables.

Author: Thanks, updated.

and Bref dependencies.

.. code-block:: none

   $ mkdir bref-mongodb-app && cd bref-mongodb-app
   $ composer init
   $ composer require bref/bref bref/extra-php-extensions mongodb/mongodb

Then initialize the serverless configuration using the ``bref`` command.

.. code-block:: none

   $ vendor/bin/bref init

This series of commands creates a file named ``index.php``. To validate your
setup, try deploying this default application. This outputs a
URL that renders a webpage with the Bref logo:

.. code-block:: none

   $ serverless deploy

Now that you have initialized the project, you will add the ``mongodb`` extension.
Locate the "Serverless config" name in the list of extensions provided by
`bref/extra-php-extension <https://github.com/brefphp/extra-php-extensions>`__.
Add it to the ``layers`` of the function in ``serverless.yaml``, this file
will look like this:

.. code-block:: yaml

   plugins:
       - ./vendor/bref/bref
       - ./vendor/bref/extra-php-extensions

   functions:
       api:
           handler: index.php
           runtime: php-83-fpm
           layers:
               - ${bref-extra:mongodb-php-83}

> Let's use the MongoDB driver with a web page that list planets from Atlas'

Reviewer 2: ```suggestion
Let's use the MongoDB driver with a web page that list planets from the Atlas
```

`sample dataset <https://www.mongodb.com/docs/atlas/sample-data/>`__.
Replace the contents of ``index.php`` with the following:

.. literalinclude:: /examples/aws-lambda/index.php
   :language: php

Redeploy the application with the new ``index.php``:

.. code-block:: none

   $ serverless deploy

The application will display an error message because the ```MONGODB_URI```
environment variable has not yet been set. We'll look at how to set this
variable in the next section.

AWS Credentials
---------------

Atlas supports passwordless authentication with AWS credentials. In any Lambda function,
AWS sets environment variables that contains the access token and secret token with
the role assigned to deployed function.

1. Open the Lambda function in the AWS console

> 2. In "Configuration > Permission", copy the "Role name"

Reviewer 2: Suggestion:
I think you could use ":guilabel:" for "Configuration > Permission".

Author: ```suggestion
2. In :guilabel:`Configuration > Permission`, copy the "Role name"
```

MongoDB Atlas documentation.

Now that the permissions have been configured, the Lambda function is allowed to access
your Atlas cluster. You can configure your application with the Atlas endpoint.

Access to Atlas clusters is also restricted by IP address. Since the range of IP that comes
from AWS is very wide, you can `allow access from everywhere <https://www.mongodb.com/docs/atlas/security/ip-access-list/>`__.

.. note::

   Using VPC Peering is recommended in order to isolate your Atlas cluster from Internet.
   This requires the Lambda function to be deployed in this AWS VPC.

Find the connection URI in the Atlas UI :guilabel:`Atlas > Deployment > Database > Connect`.
Select :guilabel:`3. AWS IAM`.
Remove the ``<AWS access key>:<AWS secret key>`` part from the URI, the credentials
will be read from environment variables.

Update the ``serverless.yml`` file to pass the environment variable ``MONGODB_URI``.

.. code-block:: yaml

   provider:
       environment:
           MONGODB_URI: "mongodb+srv://cluster0.example.mongodb.net/?authSource=%24external&authMechanism=MONGODB-AWS&retryWrites=true&w=majority"

> Finally, deploy with the new configuration. Once done you can access the function URL

Reviewer 2: ```suggestion
Finally, deploy with the new configuration. After deployment completes, you can access the function URL
```
